### DISCUSSION TITLE
Draw menu configuration per mode

### DISCUSSION THREAD

**Discussion by gennaro-tedesco** [saghen/blink.cmp#756] [Q&A]
I would like to avoid showing part of the draw menu when in cmdline mode, for instance see the below example

![SCR-20241224-kthq](https://github.com/user-attachments/assets/b76e8ea7-1528-47b8-8d51-a04195cc2d25)

where all except the command label is redundant. I have tried the following:
```lua
menu.draw.components = {
  kind = {
    text = function(ctx)
	return ctx.mode == "cmdline" and "" or ctx.kind
    end,
  },
}
```
but it seems that the condition `ctx.mode == "cmdline"` always defaults to `false`, perhaps the `ctx` is not passed in the components module? 

A similar question has been asked [here](https://github.com/Saghen/blink.cmp/issues/515), in the context of separate `winopts`; this would be the equivalent only for draw menu (perhaps it exists already and I have not found it in the docs?)

What is the current suggested approach to only draw specific UI components according to the different completion modes?

**Comment by josh-nz** [comment]
I was looking for a solution to the same problem and tried the exact same thing you did. `mode` doesn't seem to be available on the `ctx` in this function. And while if it was present this solution would work, it's also a bit of a pain since you need to adjust each component.

Separate menu opts for cmdline and editor would be a nicer, more customisable experience. But also more effort to implement.

**Comment by krovuxdev** [comment]
@gennaro-tedesco I believe you might be referring to `source_name` instead of `mode`. Have you tried using `ctx.source_name` instead?

**Reply by krovuxdev** [reply]
Oh, also  @josh-nz.

**Reply by josh-nz** [reply]
Awesome that works. I should have noticed that since I'm displaying the `source_name` in my results list. Here's what I came up with:

```lua
        draw = {
          columns = { { "label", "label_description", gap = 1 }, { "kind" }, { "source_name" }  },
          -- treesitter = { "lsp" },
          components = {
            kind = {
              text = function(ctx) return ctx.source_name == "cmdline" and "" or ctx.kind end,
            },
            source_name = {
              text = function(ctx) return ctx.source_name == "cmdline" and "" or ctx.source_name end,
            },
          },
        },
```

In the editor, completion results show all the columns, while in command mode I only see the label and label desc. There looks to be the blank columns still from the kind and source_name, but it's good enough for now. Thanks @krovuxdev

**Reply by gennaro-tedesco** [reply]
@krovuxdev Thank you for the suggestion: I am however looking at this [merged PR](https://github.com/Saghen/blink.cmp/pull/1049), which seems to be working as well.

**Reply by josh-nz** [reply]
Oh very nice, that is the better solution I was referring to. And it's part of 0.11 that released a few days ago. Will upgrade to that now and use the new features.
